### PR TITLE
Add textDocument/typeDefinition for LSP

### PR DESCRIPTION
**Pull request by andreypopp**
Doc to spec https://microsoft.github.io/language-server-protocol/specification#textDocument_typeDefinition

This works like textDocument/definition but resolves a location of a type of an expression under the cursor.

I'm not sure what to do with tsserver though.

### CODE REVIEW

**Comment by andreypopp** [comment]
This is ready for review.

File: autoload/ale/lsp/message.vim [hunk 1]
```diff
@@ -120,6 +120,15 @@ function! ale#lsp#message#Definition(buffer, line, column) abort
     \}]
 endfunction
 
+function! ale#lsp#message#TypeDefinition(buffer, line, column) abort
+    return [0, 'textDocument/typeDefinition', {
+    \   'textDocument': {
+    \       'uri': ale#path#ToURI(expand('#' . a:buffer . ':p')),
+    \   },
+    \   'position': {'line': a:line - 1, 'character': a:column - 1},
```
Comment (w0rp) [inline autoload/ale/lsp/message.vim]: You'll have to add `- 1` to the `definition` request too, if that's what we need to do here.

File: doc/ale.txt [hunk 2]
```diff
@@ -2328,6 +2343,42 @@ ALEGoToDefinitionInVSplit                           *ALEGoToDefinitionInVSplit*
...
+  definition of a type of a symbol undert the cursor. ALE will jump to a
+  definition if an LSP server provides a location to jump to. Otherwise, ALE
+  will do nothing.
+
+  You can jump back to the position you were at before going to the definition
+  of something with jump motions like CTRL-O. See |jump-motions|.
+
+  A plug mapping `<Plug>(ale_go_to_type_definition)` is defined for this
+  command.
+
```
Comment (w0rp) [inline doc/ale.txt]: Write two blank lines instead here.

**Comment by w0rp** [comment]
We can get #2224 merged first and fix the conflict here. I'll see if there's an equivalent in `tsserver`.

**Comment by w0rp** [comment]
There is indeed an equivalent for tsserver. I'll merge this now, and I'll add support for the other message later.

**Comment by w0rp** [comment]
Cheers! :beers:

**Pull request merged**